### ISSUE TITLE
Replace `H256` and `U256` with `StarkHash` and `StarkInteger`

### ISSUE THREAD

**Issue by xJonathanLEI** [xJonathanLEI/starknet-rs#13]
For convenience, the library currently uses `H256` and `U256` from the `ethereum_types` crate to represent hashes and integers on StarkNet. However, the range of these hashes and integers are actually smaller than 256 bits. Exposing public interfaces with `H256` or `U256` is thus misleading.

We should implement the `StarkHash` and `StarkInteger` types (names open to discussion) to replace the existing Ethereum types used, probably as wrappers around `H256` and `U256`, respectively, with range checks.

**Comment by milancermak** [comment]
I think they might be [felts](https://www.cairo-lang.org/docs/how_cairo_works/cairo_intro.html#field-elements), which is a 252-bit number, but doesn't quite need all the 252 bit space, as max value is 2**251 + 17 * 2**192 + 1.

**Comment by xJonathanLEI** [comment]
Yes they're felts, so we better name then `FieldElement` instead of `StarkInteger`? Anyways we still need something like that with range check.

BTW I forgot that these numbers can be negative too, so I used `U256` to represent them. Need to fix.

**Comment by milancermak** [comment]
Personally, I got used to the name `felt`, but `FieldElement` is fine with me.

I don't think the values can be negative, just in the context of Cairo, negative numbers are represented as the "upper half" of the felt's value. At least that's my understanding.

**Comment by xJonathanLEI** [comment]
I think you're right that they're only negative in the context of Cairo. Aka. they're not really two's-complement negative. I'm not entirely sure and will need to do some tests on this.

But imaging having contract bindings (we will pretty soon after implementing a signer) that return `felt`s, users would definitely expect a negative value under Cairo context to be negative in the client code too.

**Comment by xJonathanLEI** [comment]
How about this? We treat literally ALL values (including hashes and contract code and call contract results) as a single type: `UnsignedFieldElement`, which can then be converted into `SignedFieldElement` if users choose to. In this case, we'll use `SIgnedFieldElement` for return types in `abigen` binding code, for the sake of fitting into the Ciaro context.

The reason is that it seems the boundary between hashes and numbers are quite blurry in StarkWare's context. They're literally the same thing after all. So we might as well just give up on separating them.

**Comment by milancermak** [comment]
Great idea ❤️ I think that will work nicely.

**Comment by xJonathanLEI** [comment]
OK I'm starting to work on this now. The issue has become quite burning as our API surface grows.

**Comment by xJonathanLEI** [comment]
All tasks done. Closing issue.
